hwconf/hw_gesc.c

```diff
@@ -0,0 +1,492 @@
+
+    // Channel 1, 2 and 3 Configuration in PWM mode
+    TIM_OCInitStructure.TIM_OCMode = TIM_OCMode_PWM2;
+    TIM_OCInitStructure.TIM_OutputState = TIM_OutputState_Enable;
+    TIM_OCInitStructure.TIM_OutputNState = TIM_OutputNState_Enable;
+    TIM_OCInitStructure.TIM_Pulse = preface*utick;
+    TIM_OCInitStructure.TIM_OCPolarity = TIM_OCPolarity_High;
+    TIM_OCInitStructure.TIM_OCNPolarity = TIM_OCNPolarity_High;
+    TIM_OCInitStructure.TIM_OCIdleState = TIM_OCIdleState_Set;
+    TIM_OCInitStructure.TIM_OCNIdleState = TIM_OCNIdleState_Set;
+
+    TIM_OC1Init(TIM1, &TIM_OCInitStructure);
+    TIM_OC1PreloadConfig(TIM1, TIM_OCPreload_Enable);
+    TIM_OC2Init(TIM1, &TIM_OCInitStructure);
+    TIM_OC2PreloadConfig(TIM1, TIM_OCPreload_Enable);
+    TIM_OC3Init(TIM1, &TIM_OCInitStructure);
+    TIM_OC3PreloadConfig(TIM1, TIM_OCPreload_Enable);
+
+    TIM_SelectOCxM(TIM1, TIM_Channel_1, TIM_OCMode_PWM2);
+    TIM_CCxCmd(TIM1, TIM_Channel_1, TIM_CCx_Enable);
+    TIM_CCxNCmd(TIM1, TIM_Channel_1, TIM_CCxN_Enable);
+
+    TIM_SelectOCxM(TIM1, TIM_Channel_2, TIM_OCMode_Inactive);
+    TIM_CCxCmd(TIM1, TIM_Channel_2, TIM_CCx_Enable);
+    TIM_CCxNCmd(TIM1, TIM_Channel_2, TIM_CCxN_Enable);
+
+    TIM_SelectOCxM(TIM1, TIM_Channel_3, TIM_OCMode_Inactive);
+    TIM_CCxCmd(TIM1, TIM_Channel_3, TIM_CCx_Enable);
+    TIM_CCxNCmd(TIM1, TIM_Channel_3, TIM_CCxN_Enable);
+    TIM_GenerateEvent(TIM1, TIM_EventSource_COM);
+
+
+    // Automatic Output enable, Break, dead time and lock configuration
+    TIM_BDTRInitStructure.TIM_OSSRState = TIM_OSSRState_Enable;
+    TIM_BDTRInitStructure.TIM_OSSIState = TIM_OSSIState_Enable;
+    TIM_BDTRInitStructure.TIM_LOCKLevel = TIM_LOCKLevel_OFF;
+    if (deadtime < 0) {
+        TIM_BDTRInitStructure.TIM_DeadTime = conf_general_calculate_deadtime(HW_DEAD_TIME_NSEC, SYSTEM_CORE_CLOCK);
+    } else {
+        TIM_BDTRInitStructure.TIM_DeadTime = conf_general_calculate_deadtime(deadtime, SYSTEM_CORE_CLOCK);
+    }
+    TIM_BDTRInitStructure.TIM_Break = TIM_Break_Disable;
+    TIM_BDTRInitStructure.TIM_BreakPolarity = TIM_BreakPolarity_High;
+    TIM_BDTRInitStructure.TIM_AutomaticOutput = TIM_AutomaticOutput_Disable;
+    TIM_BDTRConfig(TIM1, &TIM_BDTRInitStructure);
+
+    TIM_CCPreloadControl(TIM1, ENABLE);
+    TIM_ARRPreloadConfig(TIM1, ENABLE);
+
+    TIM1->CNT = 0;
+    TIM1->EGR = TIM_EGR_UG;
+
+    TIM_SelectSlaveMode(TIM1, TIM_SlaveMode_Trigger);
+    TIM_SelectInputTrigger(TIM1, TIM_TS_ITR3);
+    TIM_SelectOnePulseMode(TIM1, TIM_OPMode_Single);
+    TIM_CtrlPWMOutputs(TIM1, ENABLE);
+
+    TIM_Cmd(TIM1, ENABLE);
+    //Timer 4 triggert Timer 1
+    TIM_Cmd(TIM4, ENABLE);
+    TIM_Cmd(TIM4, DISABLE);
+    TIM1->ARR = (breaktime+pulse2)*utick;
+    TIM1->CCR1 = breaktime*utick;
+    while ( TIM1->CNT != 0 );
+    TIM_Cmd(TIM4, ENABLE);
+
+    chThdSleepMilliseconds(1);
+    TIM_CtrlPWMOutputs(TIM1, DISABLE);
+    mc_configuration *mcconf = mempools_alloc_mcconf();
```

Comment:
You need to free the configuration when you are done, otherwise each call to this terminal command will make the memory pool run out of configurations. That is what is called a memory leak.

Reply:
argl, yes, you are right. i'm wondering why sometimes the system stops after a few tests

Reply:
Actually, the functions below expect the configuration to persist (they will store a pointer to it), so it is better to give them the pointer directly from mc_interface. There is probably a better wait to do this in general, maybe I should add something to mc_interface...

Reply:
mmh, i tried to use just a pointer without the malloc. this did not work, but i don't know anymore why.